fvm/environment/derived_data_invalidator.go

```diff
@@ -91,11 +90,32 @@ func (invalidator ProgramInvalidator) ShouldInvalidateEntries() bool {
 
 func (invalidator ProgramInvalidator) ShouldInvalidateEntry(
 	location common.AddressLocation,
-	program *interpreter.Program,
+	program *derived.Program,
 	state *state.State,
 ) bool {
-	// TODO(rbtz): switch to fine grain invalidation.
-	return invalidator.ShouldInvalidateEntries()
+	if invalidator.MeterParamOverridesUpdated {
+		// if meter parameters changed we need to invalidate all programs
+		return true
+	}
+
+	// if an account was (un)frozen we need to invalidate all
+	// programs that depend on any contract on that address.
+	for _, frozenAccount := range invalidator.FrozenAccounts {
+		_, ok := program.Dependencies[frozenAccount]
+		if ok {
+			return true
+		}
+	}
+
+	// invalidate all programs depending on any of the contracts that were updated
+	// A program has itself listed as a dependency, so that this simpler.
+	for _, key := range invalidator.ContractUpdateKeys {
+		_, ok := program.Dependencies[common.MustBytesToAddress(key.Address.Bytes())]
```

Comment:
we should precompute `common.MustBytesToAddress(key.Address.Bytes())` instead of using key directly / recomputing the same address for each program entry